test/test_polynomial.cpp

```diff
@@ -227,20 +235,50 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(test_non_integral_arithmetic_relations, T, non_int
     BOOST_CHECK_EQUAL(a * T(0.5), a / T(2));
 }
 
-BOOST_AUTO_TEST_CASE_TEMPLATE( test_self_multiply_assign, T, all_test_types )
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(test_right_shift, T, all_test_types )
 {
-    polynomial<T> a(d3a.begin(), d3a.end());
-    polynomial<T> const b(a);
-    boost::array<double, 7> const d3a_sq = {{100, -120, -44, 108, -20, -24, 9}};
-    polynomial<T> const asq(d3a_sq.begin(), d3a_sq.end());
+    polynomial<T> a(d8b.begin(), d8b.end());
+    polynomial<T> const aa(a);
+    polynomial<T> const b(d8b.begin() + 1, d8b.end());
+    polynomial<T> const c(d8b.begin() + 5, d8b.end());
+    a >>= 0u;
+    BOOST_CHECK_EQUAL(a, aa);
+    a >>= 1u;
+    BOOST_CHECK_EQUAL(a, b);
+    a = a >> 4u;
+    BOOST_CHECK_EQUAL(a, c);
+}
 
-    a *= a;
 
-    BOOST_CHECK_EQUAL(a, b*b);
-    BOOST_CHECK_EQUAL(a, asq);
+BOOST_AUTO_TEST_CASE_TEMPLATE(test_left_shift, T, all_test_types )
+{
+    polynomial<T> a(d0a.begin(), d0a.end());
+    polynomial<T> const aa(a);
+    polynomial<T> const b(d0a1.begin(), d0a1.end());
+    polynomial<T> const c(d0a5.begin(), d0a5.end());
+    a <<= 0u;
+    BOOST_CHECK_EQUAL(a, aa);    
+    a <<= 1u;
+    BOOST_CHECK_EQUAL(a, b);
+    a = a << 4u;
+    BOOST_CHECK_EQUAL(a, c);
+    polynomial<T> zero = zero_element(multiplies< polynomial<T> >());
+    // Multiplying zero by x should still be zero.
+    zero <<= 1u;
+    BOOST_CHECK_EQUAL(zero, zero_element(multiplies< polynomial<T> >()));
+}
 
-    a *= a;
 
-    BOOST_CHECK_EQUAL(a, b*b*b*b);
```

Comment:
I merged the test for self-multiplication into the general test of multiplication.